### PR TITLE
docs(throttle): first draft of docs for the throttler module

### PR DESCRIPTION
**Pull request by jmcdo29**
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] Other... Please describe:
```

## What is the current behavior?

No docs for the new `ThrottlerModule`

Issue Number: N/A

## What is the new behavior?
Docs exist for the new `ThrottlerModule`.

## Does this PR introduce a breaking change?
```
[ ] Yes
[X] No
```

## Other information
Definitely gonna need a read through and a good update from @johnbiundo, but I wanted to get a first draft up so we could figure out where to go from here.

### CODE REVIEW

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:6]: This must be `@nestjs/throttler` (missing `r` at the end).

Comment (kkoomen) [reply content/techniques/rate-limiting.md:6]: perhaps add `yarn` version as well?

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:9]: `Once *the* installation is complete, ...`

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:43]: > If you wanted to bind the guard globally, for example, you could do so but adding this provider to any module

1) Add a dot to the end of the sentence.
2) Perhaps rephrase this sentence. What about this: _Binding the guard globally can be done by adding the `ThrottlerGuard` as global provider in the main app module._

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:66]: Configuration is one of the first things people want to do and care about when reading a new page that they perhaps want to use. I suggest we move this as a sub-heading under the `Rate Limiting` section or directly below it as an H3 as well.

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:73]: `the storage setting` sounds weird to me, perhaps use `the storage provider` ?

File: content/techniques/rate-limiting.md
Comment (kkoomen) [inline content/techniques/rate-limiting.md:77]: This `set` is being cut-off incorrectly.

**Comment by kkoomen** [comment]
@jmcdo29 @johnbiundo I suggest we remove the [Rate Limiting](https://docs.nestjs.com/techniques/security#rate-limiting) section in the Security section on the website, since this documentation will replace that completely, which is also the goal of this plugin.

File: content/techniques/rate-limiting.md
Comment (kkoomen) [reply content/techniques/rate-limiting.md:73]: perhaps just rephrase it as: `the storage provider to keep track of the requests.`

**Comment by kamilmysliwiec** [comment]
@jmcdo29 we should be ready to return back to this PR :) There might be some merge conflicts now as I added a dedicated Security category.

**Comment by jmcdo29** [comment]
@kamilmysliwiec should I overwrite the `rate-limiting` section that is already in `Security`? Or should this take priority at the top and we keep the old docs at the bottom of the page, as they are still relevant?

**Comment by kkoomen** [comment]
> @kamilmysliwiec should I overwrite the `rate-limiting` section that is already in `Security`? Or should this take priority at the top and we keep the old docs at the bottom of the page, as they are still relevant?

I personally would suggest to overwrite it, because this is the new module that nest should provide and this is what users should use and read through. Getting rid of the old docs is the right choice. If things of the old docs are relevant for this version, but haven't been included yet here, then it should be added in here.

**Comment by kamilmysliwiec** [comment]
@jmcdo29 yeah feel free to overwrite the current chapter :)

**Comment by jmcdo29** [comment]
Just pulled in _way_ more than I meant to trying to clean up a conflict. I'm gonna nuke my branch and do it again to clean things back up

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:3]: A few notes:
1. I'd remove any references to "express" compatible packages
2. Since it's NestJS documentation (and `@nestjs/{..}` package), we should remove "in a more NestJS Way" as this is kinda obvious 😄 
3. Maybe we should leave the first sentence as it was just to let people know what this chapter is about (`A common technique to protect applications from brute-force attacks is **rate-limiting**.`)

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:9]:
```suggestion
Once the installation is complete, the `ThrottlerModule` can be configured as any other Nest package with `forRoot` or `forRootAsync` methods.
```

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:23]: Following other chapters, can we add a new "Async configuration" section and move this to the end?

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:64]: Btw `req` should be automatically set by the `@nestjs/graphql` package.

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:73]: Following other chapters, maybe we could wrap it in a table? Example https://docs.nestjs.com/microservices/redis#options

File: content/security/rate-limiting.md
Comment (jmcdo29) [reply content/security/rate-limiting.md:64]: `req` is there by default, but `res` needs to be added for this module, so I thought we should be explicit in adding it. [Though it's looking like Apollo adds `res` by default too](https://www.apollographql.com/docs/apollo-server/api/apollo-server/#middleware-specific-context-fields). Should we keep this, or make a note that this _should_ be default behavior?

File: content/security/rate-limiting.md
Comment (johannesschobel) [inline content/security/rate-limiting.md:25]:
```suggestion
Once the module has been imported, you can then choose how you would like to bind the `ThrottlerGuard`. Any kind of binding as mentioned in the [guards](https://docs.nestjs.com/guards) section is fine. If you wanted to bind the guard globally, for example, you could do so by adding this provider to any module
```

File: content/security/rate-limiting.md
Comment (johannesschobel) [inline content/security/rate-limiting.md:36]:
```suggestion
There may be a time where you want to bind the guard to a controller or globally, but want to disable rating limiting for one or more of your endpoints. For that, you can use the `@SkipThrottle()` decorator, to negate the throttler for an entire class or a single route. The `@SkipThrottle()` decorator can also take in a boolean for if there is a case where you want to exclude _most_ of a controller, but not every route.
```

File: content/security/rate-limiting.md
Comment (johannesschobel) [inline content/security/rate-limiting.md:71]: Wouldn't it be better to have the `async configuration` right below the regular `sync configuration`. And then all the configuration stuff? I think, this would enhance readability of this section dramatically.

**Review by johannesschobel — CHANGES_REQUESTED** [review]

File: content/security/rate-limiting.md
Comment (jmcdo29) [reply content/security/rate-limiting.md:71]: This is how most of Nest's docs are set up. If you take a look at the `Database` page you'll see a similar structure.

Comment (johannesschobel) [reply content/security/rate-limiting.md:71]: yeah, i know.. but i find this quite confusing :D in my opinion it would be better to have everything in one place.. but you're right - keeping the same structure on every page may also be a good idea :laughing:

**Comment by kamilmysliwiec** [comment]
@jmcdo29 is this ready for re-review?

**Comment by jmcdo29** [comment]
@kamilmysliwiec yes it is

**Comment by kkoomen** [comment]
@jmcdo29 There are still suggested changes from @johannesschobel that should be resolved, so I don't think it's ready for review yet.

**Comment by jmcdo29** [comment]
@kkoomen I'll make sure those get fixed before this gets merged into the docs. Just wanting to see if there are any other glaring things Kamil can think of before going through with all the changes, so they can be made all together.

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:3]: `@nestjs/throttler` here, `@nestjs/throttle` below

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [reply content/security/rate-limiting.md:25]: 👍 

+ let's add : or . by the end of the sentence

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:38]:
```suggestion
There is also the `@Throttle()` decorator which can be used to override the `limit` and `ttl` set in the global module, to give tighter or looser security options. This decorator can be used on a class or a function as well. The order for this decorator does matter, as the arguments are in the order of `limit, ttl`.
```

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:42]:
```suggestion
This module can work with websockets, but it requires some class extension. You can extend the `ThrottlerGuard` and override the `handleRequest` method like so:
```

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:72]:
```suggestion
  getRequestResponse(context: ExecutionContext) {
```

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:82]:
```suggestion
The following options are valid for the `ThrottlerModule`:
```

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [reply content/security/rate-limiting.md:103]: agree with @jmcdo29, let's stay consistent :)

File: content/security/rate-limiting.md
Comment (kamilmysliwiec) [inline content/security/rate-limiting.md:145]:
```suggestion
> info **Note** `ThrottlerStorage` can be imported from `@nestjs/throttler`.
```

File: content/security/rate-limiting.md
Comment (kkoomen) [reply content/security/rate-limiting.md]: I personally wouldn't change this part. I think `limit` `ttl` indicates a bit more clearly that these are 2 separate configurable properties.

**Comment by kamilmysliwiec** [comment]
Looks great @jmcdo29! I'm ready to merge this as long as you want me to do so

**Comment by jmcdo29** [comment]
> Looks great @jmcdo29! I'm ready to merge this as long as you want me to do so

Awesome! Let's do it! We should get the package published rather immediately then too. There was a 402 last time I tried to publish it, you can check the actions for the exact error.

**Comment by kamilmysliwiec** [comment]
@jmcdo29 I've recently updated the `NPM_TOKEN` in the Throttler repository so it should be fine. Make sure to run `npm run publish` with `--access public` flag

**Comment by jmcdo29** [comment]
Due to how semantic-release works, version 1.1.1 is published!

**Pull request merged**
